File: LC 865.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        TreeNode* ans = nullptr;
        int maxDepth = -1;
        function<int(TreeNode*, int)> dfs = [&] (TreeNode* root, int depth) -> int {
            if (root == nullptr) {
                maxDepth = max(maxDepth, depth);
                return depth;
            }
            int leftMaxDepth = dfs(root->left, depth + 1);
            int rightMaxDepth = dfs(root->right, depth + 1);
            if (leftMaxDepth == rightMaxDepth && leftMaxDepth == maxDepth) {
                ans = root;
            }
            return max(leftMaxDepth, rightMaxDepth);
        };
        dfs(root, 0);
        return ans;
    }
};

File: 1000/LC 865.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func subtreeWithAllDeepest(root *TreeNode) (ans *TreeNode) {
    mx := -1

    var dfs func(*TreeNode, int) int
    dfs = func(root *TreeNode, d int) int {
        if root == nil {
            mx = max(mx, d)
            return d
        }
        l := dfs(root.Left, d + 1)
        r := dfs(root.Right, d + 1)
        if l == r && l == mx {
            ans = root
        }
        return max(l, r)
    }
    dfs(root, 0)

    return
}